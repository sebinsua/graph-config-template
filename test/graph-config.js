const test = require('ava')

const graphConfig = require('../src')
const { using } = graphConfig

const { digraph, graph } = graphConfig()

test('can represent simple graph', t => {
  const save = graph`
    A -> B [label="LIKES"];
    A -> C [label="LOVES"];
  `

  const statements = save({
    A: { id: 10, body: 'body text' },
    B: { id: 25 },
    C: { type: ['C', 'LABEL_OF_C'], id: 100 }
  })

  t.snapshot(statements)
})

test('can represent complex graph (1-to-m, transformations)', t => {
  const A_1 = { sides: 3 }
  const B = { sides: 5 }

  const typeToLabel = props => ({ label: props.type, id: props.name })
  const save = digraph`
  A [idName=aId,${A_1}];
  B [idName=bId,${B}]
  C [idName=id,${using('C')(typeToLabel)}];
  D [idName=id,${using('D')(typeToLabel)}];

  A -> B [label="LIKES"];
  A -> C -> D [label="LOVES"];
  `

  const statements = save({
    A: { aId: 10, body: 'body text' }, // [ { aId: 10 }, { aId: 15 }, { aId: 20 } ],
    B: [{ bId: 25 }, { bId: 30 }, { bId: 35 }, { bId: 40 }, { bId: 45 }],
    C: { type: ['C', 'LABEL_OF_C'], name: 'Joe' },
    D: { type: ['D', 'LABEL_OF_D'], name: 'America' }
  })

  t.snapshot(statements)
})

test('can switch the direction of the relationships', t => {
  const save = digraph`
    A -> B [dir="forward",label="LIKES"];
    A -> C -> D [dir="back",label="LOVES"];
  `

  const statements = save({
    A: { id: 10, body: 'body text' }, // [ { aId: 10 }, { aId: 15 }, { aId: 20 } ],
    B: [{ id: 25 }, { id: 30 }, { id: 35 }, { id: 40 }, { id: 45 }],
    C: { type: ['C', 'LABEL_OF_C'], id: 100 },
    D: { type: ['D', 'LABEL_OF_D'], id: 500 }
  })

  t.snapshot(statements)
})

test('can represent nodes connected to themselves', t => {
  const save = digraph`
    opinion

    fact_1 [label="fact"]
    fact_2 [label="fact"]

    fact_1 -> fact_2
  `

  const statements = save({
    opinion: { id: 1, message: 'this is probably broken' },
    fact_1: { id: 101, message: '1 + 1 = 2' },
    fact_2: { id: 9000, message: 'snapshots are cool' }
  })

  t.snapshot(statements)
})

test('can represent some *real* database', t => {
  const save = digraph`
    parent_account [label="User"]
    child_account [label="User"]

    parent_tweet [label="Tweet"]
    child_tweet [label="Tweet"]

    parent_account -> parent_tweet [label="TWEETED"]
    child_account -> child_tweet [label="TWEETED"]

    child_tweet -> parent_tweet [label="REPLIED_TO"]
  `

  const statements = save({
    parent_account: { id: 987654321, name: 'Soul' },
    child_account: { id: 1, name: 'Creator' },
    parent_tweet: { id: 1000, message: 'Why would I do this though?' },
    child_tweet: { id: 2000, message: 'This is a bad example. I am sorry.' }
  })

  t.snapshot(statements)
})
