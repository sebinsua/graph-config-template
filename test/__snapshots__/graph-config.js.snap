exports[`can represent a complex graph with one-to-many relationships, transformations, etc 1`] = `"[{"statement":"MERGE (n:A { aId: {aId} }) SET n += $remainingProps","parameters":{"aId":10,"remainingProps":{"sides":3,"body":"body text"}}},{"statement":"MERGE (n:B { bId: {bId} }) SET n += $remainingProps","parameters":{"bId":25,"remainingProps":{"sides":5}}},{"statement":"MERGE (n:B { bId: {bId} }) SET n += $remainingProps","parameters":{"bId":30,"remainingProps":{"sides":5}}},{"statement":"MERGE (n:B { bId: {bId} }) SET n += $remainingProps","parameters":{"bId":35,"remainingProps":{"sides":5}}},{"statement":"MERGE (n:B { bId: {bId} }) SET n += $remainingProps","parameters":{"bId":40,"remainingProps":{"sides":5}}},{"statement":"MERGE (n:B { bId: {bId} }) SET n += $remainingProps","parameters":{"bId":45,"remainingProps":{"sides":5}}},{"statement":"MERGE (n:C:LABEL_OF_C { id: {id} }) SET n += $remainingProps","parameters":{"id":100,"remainingProps":{"label":"["C","LABEL_OF_C"]"}}},{"statement":"MERGE (n:D:LABEL_OF_D { id: {id} }) SET n += $remainingProps","parameters":{"id":500,"remainingProps":{"label":"["D","LABEL_OF_D"]"}}},{"statement":"MATCH (left:A { aId: {left_aId} }), (right:B { bId: {right_bId} }) MERGE (left)-[r:LIKES]-(right) SET r += $relationshipProps","parameters":{"left_aId":10,"right_bId":25,"relationshipProps":{"label":"LIKES"}}},{"statement":"MATCH (left:A { aId: {left_aId} }), (right:B { bId: {right_bId} }) MERGE (left)-[r:LIKES]-(right) SET r += $relationshipProps","parameters":{"left_aId":10,"right_bId":30,"relationshipProps":{"label":"LIKES"}}},{"statement":"MATCH (left:A { aId: {left_aId} }), (right:B { bId: {right_bId} }) MERGE (left)-[r:LIKES]-(right) SET r += $relationshipProps","parameters":{"left_aId":10,"right_bId":35,"relationshipProps":{"label":"LIKES"}}},{"statement":"MATCH (left:A { aId: {left_aId} }), (right:B { bId: {right_bId} }) MERGE (left)-[r:LIKES]-(right) SET r += $relationshipProps","parameters":{"left_aId":10,"right_bId":40,"relationshipProps":{"label":"LIKES"}}},{"statement":"MATCH (left:A { aId: {left_aId} }), (right:B { bId: {right_bId} }) MERGE (left)-[r:LIKES]-(right) SET r += $relationshipProps","parameters":{"left_aId":10,"right_bId":45,"relationshipProps":{"label":"LIKES"}}},{"statement":"MATCH (left:A { aId: {left_aId} }), (right:C:LABEL_OF_C { id: {right_id} }) MERGE (left)-[r:LOVES]-(right) SET r += $relationshipProps","parameters":{"left_aId":10,"right_id":100,"relationshipProps":{"label":"LOVES"}}},{"statement":"MATCH (left:C:LABEL_OF_C { id: {left_id} }), (right:D:LABEL_OF_D { id: {right_id} }) MERGE (left)-[r:LOVES]-(right) SET r += $relationshipProps","parameters":{"left_id":100,"right_id":500,"relationshipProps":{"label":"LOVES"}}}]"`;

exports[`can represent a simple graph 1`] = `"[{"statement":"MERGE (n:A { id: {id} }) SET n += $remainingProps","parameters":{"id":10,"remainingProps":{"body":"body text"}}},{"statement":"MERGE (n:B { id: {id} }) SET n += $remainingProps","parameters":{"id":25,"remainingProps":{}}},{"statement":"MERGE (n:C { id: {id} }) SET n += $remainingProps","parameters":{"id":100,"remainingProps":{"type":"["C","LABEL_OF_C"]"}}},{"statement":"MATCH (left:A { id: {left_id} }), (right:B { id: {right_id} }) MERGE (left)-[r:LIKES]-(right) SET r += $relationshipProps","parameters":{"left_id":10,"right_id":25,"relationshipProps":{"label":"LIKES"}}},{"statement":"MATCH (left:A { id: {left_id} }), (right:C { id: {right_id} }) MERGE (left)-[r:LOVES]-(right) SET r += $relationshipProps","parameters":{"left_id":10,"right_id":100,"relationshipProps":{"label":"LOVES"}}}]"`;

exports[`can represent complex graph (1-to-m, transformations) 1`] = `"[{\"statement\":\"MERGE (n:A { aId: {aId} }) SET n += $remainingProps\",\"parameters\":{\"aId\":10,\"remainingProps\":{\"sides\":3,\"body\":\"body text\"}}},{\"statement\":\"MERGE (n:B { bId: {bId} }) SET n += $remainingProps\",\"parameters\":{\"bId\":25,\"remainingProps\":{\"sides\":5}}},{\"statement\":\"MERGE (n:B { bId: {bId} }) SET n += $remainingProps\",\"parameters\":{\"bId\":30,\"remainingProps\":{\"sides\":5}}},{\"statement\":\"MERGE (n:B { bId: {bId} }) SET n += $remainingProps\",\"parameters\":{\"bId\":35,\"remainingProps\":{\"sides\":5}}},{\"statement\":\"MERGE (n:B { bId: {bId} }) SET n += $remainingProps\",\"parameters\":{\"bId\":40,\"remainingProps\":{\"sides\":5}}},{\"statement\":\"MERGE (n:B { bId: {bId} }) SET n += $remainingProps\",\"parameters\":{\"bId\":45,\"remainingProps\":{\"sides\":5}}},{\"statement\":\"MERGE (n:C:LABEL_OF_C { id: {id} }) SET n += $remainingProps\",\"parameters\":{\"id\":100,\"remainingProps\":{\"label\":\"[\\\"C\\\",\\\"LABEL_OF_C\\\"]\"}}},{\"statement\":\"MERGE (n:D:LABEL_OF_D { id: {id} }) SET n += $remainingProps\",\"parameters\":{\"id\":500,\"remainingProps\":{\"label\":\"[\\\"D\\\",\\\"LABEL_OF_D\\\"]\"}}},{\"statement\":\"MATCH (left:A { aId: {left_aId} }), (right:B { bId: {right_bId} }) MERGE (left)-[r:LIKES]-(right) SET r += $relationshipProps\",\"parameters\":{\"left_aId\":10,\"right_bId\":25,\"relationshipProps\":{\"label\":\"LIKES\"}}},{\"statement\":\"MATCH (left:A { aId: {left_aId} }), (right:B { bId: {right_bId} }) MERGE (left)-[r:LIKES]-(right) SET r += $relationshipProps\",\"parameters\":{\"left_aId\":10,\"right_bId\":30,\"relationshipProps\":{\"label\":\"LIKES\"}}},{\"statement\":\"MATCH (left:A { aId: {left_aId} }), (right:B { bId: {right_bId} }) MERGE (left)-[r:LIKES]-(right) SET r += $relationshipProps\",\"parameters\":{\"left_aId\":10,\"right_bId\":35,\"relationshipProps\":{\"label\":\"LIKES\"}}},{\"statement\":\"MATCH (left:A { aId: {left_aId} }), (right:B { bId: {right_bId} }) MERGE (left)-[r:LIKES]-(right) SET r += $relationshipProps\",\"parameters\":{\"left_aId\":10,\"right_bId\":40,\"relationshipProps\":{\"label\":\"LIKES\"}}},{\"statement\":\"MATCH (left:A { aId: {left_aId} }), (right:B { bId: {right_bId} }) MERGE (left)-[r:LIKES]-(right) SET r += $relationshipProps\",\"parameters\":{\"left_aId\":10,\"right_bId\":45,\"relationshipProps\":{\"label\":\"LIKES\"}}},{\"statement\":\"MATCH (left:A { aId: {left_aId} }), (right:C:LABEL_OF_C { id: {right_id} }) MERGE (left)-[r:LOVES]-(right) SET r += $relationshipProps\",\"parameters\":{\"left_aId\":10,\"right_id\":100,\"relationshipProps\":{\"label\":\"LOVES\"}}},{\"statement\":\"MATCH (left:C:LABEL_OF_C { id: {left_id} }), (right:D:LABEL_OF_D { id: {right_id} }) MERGE (left)-[r:LOVES]-(right) SET r += $relationshipProps\",\"parameters\":{\"left_id\":100,\"right_id\":500,\"relationshipProps\":{\"label\":\"LOVES\"}}}]"`;

exports[`can represent nodes connected to themselves 1`] = `"[{\"statement\":\"MERGE (n:opinion { id: {id} }) SET n += $remainingProps\",\"parameters\":{\"id\":1,\"remainingProps\":{\"message\":\"this is probably broken\"}}},{\"statement\":\"MERGE (n:fact { id: {id} }) SET n += $remainingProps\",\"parameters\":{\"id\":101,\"remainingProps\":{\"label\":\"fact\",\"message\":\"1 + 1 = 2\"}}},{\"statement\":\"MERGE (n:fact { id: {id} }) SET n += $remainingProps\",\"parameters\":{\"id\":9000,\"remainingProps\":{\"label\":\"fact\",\"message\":\"snapshots are cool\"}}},{\"statement\":\"MATCH (left:fact { id: {left_id} }), (right:fact { id: {right_id} }) MERGE (left)-[r:CONNECTED_TO]-(right) SET r += $relationshipProps\",\"parameters\":{\"left_id\":101,\"right_id\":9000,\"relationshipProps\":{}}}]"`;

exports[`can represent simple graph 1`] = `"[{\"statement\":\"MERGE (n:A { id: {id} }) SET n += $remainingProps\",\"parameters\":{\"id\":10,\"remainingProps\":{\"body\":\"body text\"}}},{\"statement\":\"MERGE (n:B { id: {id} }) SET n += $remainingProps\",\"parameters\":{\"id\":25,\"remainingProps\":{}}},{\"statement\":\"MERGE (n:C { id: {id} }) SET n += $remainingProps\",\"parameters\":{\"id\":100,\"remainingProps\":{\"type\":\"[\\\"C\\\",\\\"LABEL_OF_C\\\"]\"}}},{\"statement\":\"MATCH (left:A { id: {left_id} }), (right:B { id: {right_id} }) MERGE (left)-[r:LIKES]-(right) SET r += $relationshipProps\",\"parameters\":{\"left_id\":10,\"right_id\":25,\"relationshipProps\":{\"label\":\"LIKES\"}}},{\"statement\":\"MATCH (left:A { id: {left_id} }), (right:C { id: {right_id} }) MERGE (left)-[r:LOVES]-(right) SET r += $relationshipProps\",\"parameters\":{\"left_id\":10,\"right_id\":100,\"relationshipProps\":{\"label\":\"LOVES\"}}}]"`;

exports[`can switch the direction of the relationships 1`] = `"[{\"statement\":\"MERGE (n:A { id: {id} }) SET n += $remainingProps\",\"parameters\":{\"id\":10,\"remainingProps\":{\"body\":\"body text\"}}},{\"statement\":\"MERGE (n:B { id: {id} }) SET n += $remainingProps\",\"parameters\":{\"id\":25,\"remainingProps\":{}}},{\"statement\":\"MERGE (n:B { id: {id} }) SET n += $remainingProps\",\"parameters\":{\"id\":30,\"remainingProps\":{}}},{\"statement\":\"MERGE (n:B { id: {id} }) SET n += $remainingProps\",\"parameters\":{\"id\":35,\"remainingProps\":{}}},{\"statement\":\"MERGE (n:B { id: {id} }) SET n += $remainingProps\",\"parameters\":{\"id\":40,\"remainingProps\":{}}},{\"statement\":\"MERGE (n:B { id: {id} }) SET n += $remainingProps\",\"parameters\":{\"id\":45,\"remainingProps\":{}}},{\"statement\":\"MERGE (n:C { id: {id} }) SET n += $remainingProps\",\"parameters\":{\"id\":100,\"remainingProps\":{\"type\":\"[\\\"C\\\",\\\"LABEL_OF_C\\\"]\"}}},{\"statement\":\"MERGE (n:D { id: {id} }) SET n += $remainingProps\",\"parameters\":{\"id\":500,\"remainingProps\":{\"type\":\"[\\\"D\\\",\\\"LABEL_OF_D\\\"]\"}}},{\"statement\":\"MATCH (left:A { id: {left_id} }), (right:B { id: {right_id} }) MERGE (left)-[r:LIKES]->(right) SET r += $relationshipProps\",\"parameters\":{\"left_id\":10,\"right_id\":25,\"relationshipProps\":{\"dir\":\"forward\",\"label\":\"LIKES\"}}},{\"statement\":\"MATCH (left:A { id: {left_id} }), (right:B { id: {right_id} }) MERGE (left)-[r:LIKES]->(right) SET r += $relationshipProps\",\"parameters\":{\"left_id\":10,\"right_id\":30,\"relationshipProps\":{\"dir\":\"forward\",\"label\":\"LIKES\"}}},{\"statement\":\"MATCH (left:A { id: {left_id} }), (right:B { id: {right_id} }) MERGE (left)-[r:LIKES]->(right) SET r += $relationshipProps\",\"parameters\":{\"left_id\":10,\"right_id\":35,\"relationshipProps\":{\"dir\":\"forward\",\"label\":\"LIKES\"}}},{\"statement\":\"MATCH (left:A { id: {left_id} }), (right:B { id: {right_id} }) MERGE (left)-[r:LIKES]->(right) SET r += $relationshipProps\",\"parameters\":{\"left_id\":10,\"right_id\":40,\"relationshipProps\":{\"dir\":\"forward\",\"label\":\"LIKES\"}}},{\"statement\":\"MATCH (left:A { id: {left_id} }), (right:B { id: {right_id} }) MERGE (left)-[r:LIKES]->(right) SET r += $relationshipProps\",\"parameters\":{\"left_id\":10,\"right_id\":45,\"relationshipProps\":{\"dir\":\"forward\",\"label\":\"LIKES\"}}},{\"statement\":\"MATCH (left:A { id: {left_id} }), (right:C { id: {right_id} }) MERGE (left)<-[r:LOVES]-(right) SET r += $relationshipProps\",\"parameters\":{\"left_id\":10,\"right_id\":100,\"relationshipProps\":{\"dir\":\"back\",\"label\":\"LOVES\"}}},{\"statement\":\"MATCH (left:C { id: {left_id} }), (right:D { id: {right_id} }) MERGE (left)<-[r:LOVES]-(right) SET r += $relationshipProps\",\"parameters\":{\"left_id\":100,\"right_id\":500,\"relationshipProps\":{\"dir\":\"back\",\"label\":\"LOVES\"}}}]"`;
